### PR TITLE
Improve performance on the cloud provider node-controller

### PR DESCRIPTION
**Pull request by aojea**
Following up on https://github.com/kubernetes/kubernetes/pull/113104 , use the new option to parallelize the reconcile loop that updates the nodes addresses and labels.

This is a problem in large clusters, since this loop is the one that reconciles the nodes addresses, it may cause that some instance get assigned a public ip but it takes a long time to represent it in the Node.Status
/kind cleanup

```release-note
NONE
```

```
go test -timeout 30s -run ^TestUpdateNodeStatus$ k8s.io/cloud-provider/controllers/node -v -race
=== RUN   TestUpdateNodeStatus
=== RUN   TestUpdateNodeStatus/single_thread
    node_controller_test.go:2537: 1 workers: processed 100 nodes int 1.055595262s 
=== RUN   TestUpdateNodeStatus/5_workers
    node_controller_test.go:2537: 5 workers: processed 100 nodes int 216.990972ms 
=== RUN   TestUpdateNodeStatus/10_workers
    node_controller_test.go:2537: 10 workers: processed 100 nodes int 112.422435ms 
=== RUN   TestUpdateNodeStatus/30_workers
    node_controller_test.go:2537: 30 workers: processed 100 nodes int 46.243204ms 
--- PASS: TestUpdateNodeStatus (1.44s)
    --- PASS: TestUpdateNodeStatus/single_thread (1.06s)
    --- PASS: TestUpdateNodeStatus/5_workers (0.22s)
    --- PASS: TestUpdateNodeStatus/10_workers (0.11s)
    --- PASS: TestUpdateNodeStatus/30_workers (0.05s)
```

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 1]
```diff
@@ -172,15 +173,15 @@ func (cnc *CloudNodeController) Run(stopCh <-chan struct{}, controllerManagerMet
...
-	// very infrequently. DO NOT MODIFY this to perform frequent operations.
+	// The periodic loop for updateNodeStatus polls the Cloud Provider periodically
+	// to reconcile the nodes addresses and labels.
 	go wait.Until(func() {
 		if err := cnc.UpdateNodeStatus(context.TODO()); err != nil {
 			klog.Errorf("failed to update node status: %v", err)
 		}
 	}, cnc.nodeStatusUpdateFrequency, stopCh)
 
+	// This is only used during bootstrapping
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:184]: Please correct me if I'm wrong, but I don't think it's really true.
This is invoke on any node addition to the cluster, which may happen outside of bootstrapping.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 2]
```diff
@@ -251,28 +252,34 @@ func (cnc *CloudNodeController) syncHandler(key string) error {
...
 			klog.Errorf("Error getting instance metadata for node addresses: %v", err)
-			continue
+			return
 		}
-		cnc.updateNodeAddress(ctx, &nodes.Items[i], instanceMetadata)
-	}
 
-	for _, node := range nodes.Items {
+		cnc.updateNodeAddress(ctx, node, instanceMetadata)
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:275]: FTR - I checked the getInstanceNodeAddresses, updateNodeAddresses and reconcileNodeLabels for thread-safety and we're fine here.

File: staging/src/k8s.io/cloud-provider/fake/fake.go [hunk 3]
```diff
@@ -106,11 +106,10 @@ type Route struct {
 }
 
 func (f *Cloud) addCall(desc string) {
-	f.addCallLock.Lock()
-	defer f.addCallLock.Unlock()
-
 	time.Sleep(f.RequestDelay)
 
+	f.addCallLock.Lock()
+	defer f.addCallLock.Unlock()
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/fake/fake.go:112]: Good catch!

**Review by wojtek-t — COMMENTED** [review]
@aojea - thanks for doing this!

Just one minor comment - PTAL

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 1 above)
Comment (aojea) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:184]: yes, let me add a better comment, this is for new added nodes that are tainined

File: staging/src/k8s.io/cloud-provider/fake/fake.go (hunk 3 above)
Comment (aojea) [reply staging/src/k8s.io/cloud-provider/fake/fake.go:112]: haha, now I think it was fun, but it drove me crazy, adding more workers to the test didn't give any improvement

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 4]
```diff
@@ -251,28 +253,40 @@ func (cnc *CloudNodeController) syncHandler(key string) error {
...
-	for i := range nodes.Items {
-		instanceMetadata, err := cnc.getInstanceNodeAddresses(ctx, &nodes.Items[i])
+	updateNodeFunc := func(piece int) {
+		node := nodes[piece].DeepCopy()
+		// Do not process nodes that are still tainted, those will be processed by syncNode()
+		cloudTaint := getCloudTaint(node.Spec.Taints)
+		if cloudTaint != nil {
+			klog.V(5).Infof("This node %s is still tainted. Will not process.", node.Name)
+			return
+		}
```
Comment (aojea) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:273]: This is also in `updateNodeAddress`, but this way we avoid both reconcile loops working in parallel on the same objects and reduce the number of API calls to the cloud provider

**Comment by aojea** [comment]
> pod "downward-api-8578cf00-323c-43d6-aad2-bc1fe9364f07" failed with status: {Phase:Failed Conditions:[] Message:Pod was rejected: Node didn't have enough resource: cpu, requested: 250, used: 760, capacity: 1000 Reason:OutOfcpu NominatedNodeName: HostIP: PodIP: PodIPs:[] StartTime:2023-02-06 09:21:40 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[] QOSClass: EphemeralContainerStatuses:[]}
    <*errors.errorString | 0xc001465cd0>: {

@bobbypage @SergeyKanzhelev rings a bell? https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115521/pull-kubernetes-node-e2e-containerd/1622522250320154624/



/test pull-kubernetes-node-e2e-containerd

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**